### PR TITLE
Python mesh coordinates reorder

### LINKED ISSUE #2720: Shuffle array shapes returned from mesh classes
Based on discussion with @paulromano and @eepeterson:

The shape of vertices, and centroids returned from our mesh classes is currently `(ndim, dim0, ..., dimn)`. While this is convenient for unpacking x, y, and z coordinates in a way that is useful for some plotting operations, 

```python
xx, yy, zz = mesh.vertices
```

it makes what indexing a centroid or vertex by an `ijk` tuple (i.e. `(2, 3, 6)`  counterintuitive

```python
centroid = mesh.centroids[(:, *ijk)]
```

These properties of the mesh class will be updated to a shape of `(dim0, ..., dimn, ndim)` to allow indexing to occur more naturally:

```python
centroid = mesh.centroids[ijk]
```

Note that x, y, z values can still be unpacked fairly easily with the following command:

```python
xx, yy, zz = np.rollaxis(mesh.vertices, -1)
```

### PR DESCRIPTION
**Pull request by pshriwise**
# Description

I'm not totally happy with the necessary updates to VTK writing, but as any changes there would result in changes to the reference VTK files in the test suite, I'm going to leave that to a future PR.

Fixes #2720

# Checklist

- [x] I have performed a self-review of my own code
~~- [ ] I have run [clang-format](https://docs.openmc.org/en/latest/devguide/styleguide.html#automatic-formatting) on any C++ source files (if applicable)~~
- [x] I have followed the [style guidelines](https://docs.openmc.org/en/latest/devguide/styleguide.html#python) for Python source files (if applicable)
- [x] I have made corresponding changes to the documentation (if applicable)
- [x] I have added tests that prove my fix is effective or that my feature works (if applicable)

### CODE REVIEW

File: openmc/mesh.py [hunk 1]
```diff
@@ -172,25 +172,26 @@ def _grids(self):
...
+            be unpacked along the first dimension with xx, yy, zz =
+            np.rollaxis(mesh.vertices, -1).
 
         """
         return self._generate_vertices(*self._grids)
 
     @staticmethod
     def _generate_vertices(i_grid, j_grid, k_grid):
-        """Returns an array with shape (3, i_grid.size+1, j_grid.size+1, k_grid.size+1)
+        """Returns an array with shape (i_grid.size+1, j_grid.size+1, k_grid.size+1, 3)
```
Comment (eepeterson) [inline openmc/mesh.py:191]:
```suggestion
        """Returns an array with shape (i_grid.size, j_grid.size, k_grid.size, 3)
```

**Review by eepeterson — CHANGES_REQUESTED** [review]
Looking through these changes and reviewing the necessary fix in PR #2736 I think I'd like to see the following changes to the mesh classes:

1. A single `get_vertices()` method that by default returns vertices in cartesian coordinates, but can be used to also return the native coordinates.
2. A single `get_centroids()` method that by default returns vertices in cartesian coordinates, but can be used to also return the native coordinates.
3. Rather than having staticmethods floating around the different mesh classes I think maybe make the transformation functions module level functions or put them in a separate file. 

The fact that we have a number of static functions that don't require any class info (`_generate_vertices`, `_convert_to_cartesian`) feels a bit wrong to me. I'm open to hearing other opinions, but I feel like the non-uniformity of the code base / API is out-weighing the convenience of attribute-style accessors for `vertices` and `centroids`.

File: openmc/mesh.py [hunk 2]
```diff
@@ -1840,13 +1841,13 @@ def _convert_to_cartesian(arr, origin: Sequence[float]):
         """Converts an array with xyz values in the first dimension (shape (3, ...))
```
[the remark this answers is not in the document]
Comment (pshriwise) [reply openmc/mesh.py:1841]: Thanks! I've just cleaned that up along with other similar patterns.

**Comment by pshriwise** [comment]
> Looking through these changes and reviewing the necessary fix in PR #2736 I think I'd like to see the following changes to the mesh classes:
> 
> 1. A single `get_vertices()` method that by default returns vertices in cartesian coordinates, but can be used to also return the native coordinates.
> 2. A single `get_centroids()` method that by default returns vertices in cartesian coordinates, but can be used to also return the native coordinates.

I can see how this would help with the fix in #2736 would be handled more cleanly if we have methods like these. The esoteric call to `StructuredMesh.vertices.fget(self)` could become `self.get_vertices(native=True)` (or whatever we decide the Cartesian/Mesh coordinates toggle should be called), which is much more clear.

My concern is the level of indirection in the implementation as opposed to the API itself. In the absence of that complication, I tend to prefer properties, but I think our goals align well enough to make this transition happen 😄.

> 3. Rather than having staticmethods floating around the different mesh classes I think maybe make the transformation functions module level functions or put them in a separate file.
> The fact that we have a number of static functions that don't require any class info (`_generate_vertices`, `_convert_to_cartesian`) feels a bit wrong to me. I'm open to hearing other opinions, but I feel like the non-uniformity of the code base / API is out-weighing the convenience of attribute-style accessors for `vertices` and `centroids`.

To me, `_convert_to_cartesian` fits the purpose of a static method quite well. It provides functionality within the context of the class that it is attached to and allows that method to be applied within the base class in methods like `write_data_to_vtk` without knowledge of the subclass type with implementation determined by the type of the class instance `self` in those methods. Were we to make those methods module-level functions, we'd need to a) differentiate them by name (something like `_convert_cylindrical_to_cartesian` and `_convert_spherical_to_cartesian`) and then identify which class type to use in class methods with a pattern like:

```python
if isinstance(self, CylindricalMesh):
    coords = _convert_cylindrical_to_cartesian(coords)
elif isinstance(self, SphericalMesh):
    coords = _convert_spherical_to_cartesian(coords)
```

What we have currently seems more elegant to me. Unless you have something else in mind?

`_generate_vertices` is somewhat different. Making that a module-level function wouldn't induce the same "`isinstance`" patterns as the coordinate conversion methods, but still, I think the fact that it implements functionality specific to the mesh class is enough to keep it tied to the `StructuredMesh` class definition itself.

File: tests/unit_tests/test_mesh.py
[the remark this answers is not in the document]
Comment (pshriwise) [reply tests/unit_tests/test_mesh.py:1]: All clean ✨

**Comment by pshriwise** [comment]
> My concern is the level of indirection in the implementation as opposed to the API itself. In the absence of that complication, I tend to prefer properties, but I think our goals align well enough to make this transition happen 😄.
> 
> >

@eepeterson I'll plan on this transition in a separate PR to keep this one focused on the reordering.

File: openmc/mesh.py (hunk 1 above)
Comment (eepeterson) [reply openmc/mesh.py:190]: @pshriwise the call to `meshgrid` doesn't add 1 to the size of each array so the docstring is incorrect

File: openmc/mesh.py [hunk 3]
```diff
@@ -1543,14 +1540,15 @@ def centroids_cylindrical(self):
 
     @staticmethod
     def _convert_to_cartesian(arr, origin: Sequence[float]):
-        """Converts an array with xyz values in the first dimension (shape (3, ...))
+        """Converts an array with xyz values in the last dimension (shape (..., 3))
```
Comment (eepeterson) [inline openmc/mesh.py:1543]:
```suggestion
        """Converts an array with r, phi, z values in the last dimension (shape (..., 3))
```

File: openmc/mesh.py [hunk 4]
```diff
@@ -1839,16 +1837,16 @@ def centroids_spherical(self):
...
```
Comment (eepeterson) [inline openmc/mesh.py:1840]:
```suggestion
        """Converts an array with r, theta, phi values in the last dimension (shape (..., 3))
```

File: openmc/mesh.py (hunk 1 above)
Comment (pshriwise) [reply openmc/mesh.py:190]: Thanks! I didn't catch that when reordering these.

**Review by eepeterson — CHANGES_REQUESTED** [review]
Just some corrections in docstrings then I think this is good to go

**Comment by pshriwise** [comment]
@eepeterson I realized that, even though they're flat, the equivalent unstructured mesh properties should be consistent with the structured mesh classes and have xyz coordinates at the end. This made for some nice cleanup in that class IMO.

File: openmc/mesh.py [hunk 5]
```diff
@@ -1543,14 +1540,15 @@ def centroids_cylindrical(self):
...
-        """Converts an array with xyz values in the first dimension (shape (3, ...))
+        """Converts an array with r, phi, z values in the last dimension (shape (..., 3))
         to Cartesian coordinates.
         """
-        x = arr[0, ...] * np.cos(arr[1, ...]) + origin[0]
-        y = arr[0, ...] * np.sin(arr[1, ...]) + origin[1]
-        arr[0, ...] = x
-        arr[1, ...] = y
-        arr[2, ...] += origin[2]
+        print(arr.shape)
```
Comment (eepeterson) [inline openmc/mesh.py:1546]:
```suggestion
```
Need to remove print statement

**Review by eepeterson — CHANGES_REQUESTED** [review]
one print statement left over otherwise good to go

**Review by eepeterson — APPROVED** [review]
Thanks @pshriwise!

**Pull request merged**